src/Client/Connection.cpp

```diff
@@ -232,12 +232,27 @@ void Connection::disconnect()
     maybe_compressed_out = nullptr;
     in = nullptr;
     last_input_packet_type.reset();
-    out = nullptr; // can write to socket
+    std::exception_ptr finalize_exception;
+    try
+    {
+        // finalize() can write to socket and throw an exception.
+        out->finalize();
+    }
+    catch (...)
+    {
+        /// Don't throw an exception here, it will leave Connection in invalid state.
+        finalize_exception = std::current_exception();
+    }
+    out = nullptr;
```

Comment:
Maybe better to wrap `socket` with some RAII structure?

Reply:
I don't see how it will help. Can you share more what do you mean?

Reply:
Instead of writing try..catch you can have a class that will call `close()` in the destructor. Or maybe since we have Poco code in the base now, we can add `close()` to the socket destructor.

Reply:
But there is not problem with socket. The problem is with `WriteBufferToPocoSocket`. It calls `finalize()` in destructor and can throw an exception (so we can get terminate because of it). So, I call finalize() explicitly. And another problem is that in case of exception in `out->finalize();` we can leave `Connection` in invalid state (because after `out = nullptr;` there is still some work/cleaning, not only closing socket), so we should catch exception from `finalize` and throw it later. I don't see how we can avoid try/catch here.

Reply:
Why don't you want to handle this exception in `~WriteBufferFromPocoSocket`? It's used in other places and can potentially cause termination there.

Reply:
I just don't like that this solution won't cover all the cases when `WriteBufferFromPocoSocket` is used. Handling the exception inside of `WriteBufferFromPocoSocket ` would be a better general solution.